### PR TITLE
Reduce Location allocations while running analyzers.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/23427

### CODE REVIEW

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 1]
```diff
@@ -613,7 +614,7 @@ private bool IsInGeneratedCode(Location location, Compilation compilation, Cance
             }
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: this is safe due to a check on line 610 that location.IsInSource

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 2]
```diff
@@ -613,7 +614,7 @@ private bool IsInGeneratedCode(Location location, Compilation compilation, Cance
...
 
             // Check if this is a generated code location.
-            if (IsGeneratedOrHiddenCodeLocation(location))
+            if (IsGeneratedOrHiddenCodeLocation(location.SourceTree, location.SourceSpan))
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: this is safe due to a check on line 610 that location.IsInSource

**Comment by CyrusNajmabadi** [comment]
Tagging @sharwell @Therzok

**Comment by CyrusNajmabadi** [comment]
Also tagging @dotnet/roslyn-compiler @mavasani as this hits analyzers.

**Comment by CyrusNajmabadi** [comment]
This saves 8m allocations and 166MB of garbage.

**Comment by sharwell** [comment]
@Pilchie for ask mode

**Comment by sharwell** [comment]
I'll get the ask mode template added tonight.

**Comment by sharwell** [comment]
### Customer scenario

Running diagnostic analyzers in the IDE is slower than desired.

### Bugs this fixes

Fixes #23427

### Workarounds, if any

None needed.

### Risk

Low. The type responsible for the allocations was a simple type, and passing the fields instead of the wrapper object met the needs of the hot path.

### Performance impact

Improves performance by reducing allocation rates on a known hot path.

### Is this a regression from a previous update?

No.

### Root cause analysis

The original code was simple; no special accommodations were required prior to profiling.

### How was the bug found?

Profiling.

### Test documentation updated?

N/A

**Comment by sharwell** [comment]
@Pilchie the template is ready now

**Comment by Pilchie** [comment]
Thanks - approved.

**Pull request merged**